.github/workflows/lint-test.yaml

```diff
@@ -0,0 +1,51 @@
+name: Lint and Test Charts
+
+on:
+  push:
+    branches-ignore:
+    - 'main'
+    - 'repository'
+    
+  pull_request:
+    branches:
+    - '**'
+
+jobs:
+  lint-test:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Install Helm
+        uses: azure/setup-helm@v1
+        with:
+          version: v3.4.1
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
+
+      - name: Install chart-testing
+        uses: helm/chart-testing-action@v2.1.0
+
+#      - name: Run chart-testing (list-changed)
+#        id: list-changed
+#        run: |
+#          changed=$(ct list-changed --config ct.yaml)
+#          if [[ -n "$changed" ]]; then
+#            echo "::set-output name=changed::true"
+#          fi
```

Comment:
I assume something is wrong with this because of helm 3? I found a similar problem on other repos.

Reply:
The issue is not with helm 3 but with the fact that the GitHub runner only has 2 CPUs and 8 GB memory, which causes a failure with the default value of `replicas: 3`.  

The attached [log](https://github.com/opensearch-project/helm-charts/files/7171243/ct-install-log.txt) shows that `ct install` succeeds in testing both charts when using Minikube on my local machine with enough resources.

I can confirm that using `replicas: 1` in the values file for the `opensearch` chart within the GitHub runner succeeds.

The Kind GitHub action is fairly limited in what we can do with it and so I switched to the Minikube GitHub action. Setting `--cpus 2 --memory 6144mb`.  What fails is a replica count greater than 1 **but only within the GitHub runner**.

I do agree that we want to test and not only lint. I propose we (1) keep the `replicas: 3` in the values file and (2) override it to `1` in the GitHub actions workflow, which does provide some level of testing. 

What are your thoughts?

Reply:
After a bit of testing various options, we have a few options:

**Option 1:** Leave `replicas: 3` and use an external GitHub runner. 

I don't really like this option.

**Option 2:** Set `replicas: 1` and include a note indicating the value should not be set to a value of less than 3 for production operational environments.

This is also a less than ideal option because, by default, this results in a single copy of the index. So folks would almost always need to override this value. But this will allow tests to pass.

The reason why this option is 'less than ideal' is because OpenSearch sets the index replica count to 1 and there's a rule that disallows the index replica to be deployed to the same pod (a.k.a. OpenSearch node).

**Option 3:** Leave `replicas: 3` and 'fix-up' the `charts/opensearch/values.yaml` file **before** running chart testing.

I think this is our best option as it tests the chart and keeps minimum operating values 'as is'. In the end, we aren't testing OpenSearch but are testing the chart.

Reply:
For the curious, [here](https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true) is a successful run with linting and testing.  I simply used `sed` in a custom action to modify the values file.

Reply:
We have tried option 1 back in ODFE and it is painful to maintain a provision script, bootstrap to GitHub code, and the cleanup code. Sometimes the cleanup is not working and we have to manually delete the runner in GitHub.

I like Option 3 and agree with your comments. We are testing if charts work.

Reply:
By fix-up, we mean override with something like `--set replicas=2`? 

I don't see it in the step: https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true#step:10:30 

Maybe I'm misunderstanding. Sorry.

Reply:
you can override the defaults in CI by creating a `ci/ci-values.yaml` file inside the chart.

> Charts may have multiple custom values files matching the glob pattern '*-values.yaml' in a directory named 'ci' in the root of the chart's directory. The chart is installed and tested for each of these files. If no custom values file is present, the chart is installed and tested with defaults.

From https://github.com/helm/chart-testing/blob/main/doc/ct_install.md

Reply:
I'd recommend putting values in here to scale CI to be as small and fast as possible: no persistence, single replica etc.

Reply:
> The Kind GitHub action is fairly limited in what we can do with it and so I switched to the Minikube GitHub action.

I'm quite familiar with `kind` and the github action. What issues did you run into here?

The nice thing about `kind` is the speed compared to `minikube`.

Reply:
@smlx Thanks for pointing out the `ci` folder. I didn't see that option when reviewing the `ct install` command. I do think this is the best option.

Regarding the use of Minikube over Kind, I should have clarified that I switched to Minikube in an attempt to use all resources (cpu and memory) available on the GitHub runner by adding the following step:

```
# See https://github.com/marketplace/actions/setup-minikube-kubernetes-cluster
- name: Create minikube cluster
  uses: manusa/actions-setup-minikube@v2.4.2
  if: steps.list-changed.outputs.changed == 'true'
  with:
    minikube version: 'v1.23.0'
    kubernetes version: 'v1.21.4'
    driver: docker
    github token: ${{ secrets.GITHUB_TOKEN }}
    start args: '--addons metrics-server --cpus 2 --memory 6144mb'
```

This isn't needed anymore since `replicas: 1` will be set in the CI-specific values file.

Reply:
Hi @DandyDeveloper 

> By fix-up, we mean override with something like `--set replicas=2`?
> 
> I don't see it in the step: https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true#step:10:30
> 
> Maybe I'm misunderstanding. Sorry.

I had created the following step before running `ct install`:

```
- name: Fixup values file
  run: |
    sed -i 's/replicas: 3/replicas: 1/g' charts/opensearch/values.yaml
```

Based on what @smlx pointed out, this step of course isn't needed in lieu of the CI-specific values file.

Reply:
Hi @smlx 

> The nice thing about `kind` is the speed compared to `minikube`.

FWIW, the duration difference between Minikube and Kind for their respective steps was marginal:

* Kind: [1 min 49 seconds](https://github.com/mprimeaux/helm-charts/runs/3621760978?check_suite_focus=true)
* Minikube: [1 min 14 seconds](https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true)

Nevertheless, as suggested, I reverted to using Kind.

Reply:
I also kinda agree on using `kind` only. Thanks @mprimeaux for this.

Reply:
Kind +1